cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -116,6 +117,125 @@ TEST(TestScalarNested, ListElementInvalid) {
+
+      args.start = 1;
+      expected =
+          ArrayFromJSON(fixed_size_list(value_type, 1), "[[2], [5], [null], null]");
+      CheckScalarUnary("list_slice", input, expected, &args);
+
+      args.start = 2;
+      args.stop = 4;
+      expected = ArrayFromJSON(fixed_size_list(value_type, 2),
+                               "[[3, null], [null, null], [null, null], null]");
+      CheckScalarUnary("list_slice", input, expected, &args);
+    }
+  }
+}
+
+TEST(TestScalarNested, ListSliceOutputEqualsInputType) {
+  // Default is to return same type as the one passed in.
+  auto inputs = {
+      ArrayFromJSON(list(int8()), "[[1, 2, 3], [4, 5], [6, null], null]"),
+      ArrayFromJSON(large_list(int8()), "[[1, 2, 3], [4, 5], [6, null], null]"),
+      ArrayFromJSON(fixed_size_list(int8(), 2), "[[1, 2], [4, 5], [6, null], null]")};
+  for (auto input : inputs) {
+    ListSliceOptions args(/*start=*/0, /*stop=*/2, /*step=*/1);
+    auto expected = ArrayFromJSON(input->type(), "[[1, 2], [4, 5], [6, null], null]");
+    CheckScalarUnary("list_slice", input, expected, &args);
+  }
+}
+
+TEST(TestScalarNested, ListSliceBadParameters) {
+  auto input = ArrayFromJSON(list(int32()), "[[1]]");
+
+  // negative start
+  ListSliceOptions args(/*start=*/-1, /*stop=*/1, /*step=*/1,
+                        /*return_fixed_size_list=*/true);
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "`start`(-1) should be greater than 0 and smaller than `stop`(1)"),
+      CallFunction("list_slice", {input}, &args));
+  // start greater than stop
+  args.start = 1;
+  args.stop = 0;
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "`start`(1) should be greater than 0 and smaller than `stop`(0)"),
+      CallFunction("list_slice", {input}, &args));
+  // start same as stop
+  args.stop = args.start;
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "`start`(1) should be greater than 0 and smaller than `stop`(1)"),
+      CallFunction("list_slice", {input}, &args));
+  // stop not set and FixedSizeList requested
+  args.stop = std::nullopt;
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      NotImplemented,
+      ::testing::HasSubstr("NotImplemented: Unable to produce FixedSizeListArray without "
+                           "`stop` being set."),
+      CallFunction("list_slice", {input}, &args));
+  // stop not set and ListArray requested
+  args.stop = std::nullopt;
+  args.return_fixed_size_list = false;
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      NotImplemented, ::testing::HasSubstr("Slicing to end not yet implemented"),
+      CallFunction("list_slice", {input}, &args));
+  // step other than `1` not implmented
+  args.stop = 2;
+  args.step = 2;
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      NotImplemented,
+      ::testing::HasSubstr("Setting `step` to anything other than 1 is not supported"),
+      CallFunction("list_slice", {input}, &args));
+}
+
```

Comment:
I'd still like to see what happens when the child array has its own offset.

Reply:
Ah, sorry. Added in https://github.com/apache/arrow/pull/14395/commits/6745576308b5fd30acacb4cebc80d89344a0f8d1, let me know if that should be adjusted in any way.

Reply:
@milesgranger I think what you added is still a "normal" slice, where the parent list array is sliced. 

A python snippet (but I suppose something similar can be done here in C++), but so you can create a ListArray where the child array itself has an offset separately from the parent list array:

```python
>>> arr = pa.array([[1], [2, 3], None, [4]])
>>> arr.offsets.tolist()
[0, 1, 3, 3, 4]
>>> arr.values.tolist()
[1, 2, 3, 4]

# slicing a list array slices the offsets, but keeps the child array intact (not sliced)
>>> arr2 = arr.slice(2) 
>>> arr2.offset
2
>>> arr2.offsets.offset
2
>>> arr2.offsets.tolist()
[3, 3, 4]
>>> arr2.values.offset
0
>>> arr2.values.tolist()
[1, 2, 3, 4]

# manually creating a non-sliced ListArray (i.e. offsets are not sliced) but with child array with offset 
>>> arr3 = pa.ListArray.from_arrays(pa.array([0, 1, 3]), pa.array([0, 1, 2, 3, 4]).slice(2))
>>> arr3
<pyarrow.lib.ListArray object at 0x7f0aed532ec0>
[
  [
    2
  ],
  [
    3,
    4
  ]
]

>>> arr3.offset
0
>>> arr3.values.tolist()
[2, 3, 4]
>>> arr3.values.offset
2
```


@lidavidm is this what you meant? (since the basic sliced arrays are already covered by `CheckScalarUnary`)

Reply:
:man_facepalming: but of course, that makes more sense. Thank you for that. Updated in https://github.com/apache/arrow/pull/14395/commits/dcff4d1e8e098d3fe47f7de6ecf1f1eae65b210b